### PR TITLE
enable/disable allow list

### PR DESCRIPTION
**Pull request by patrick-ogrady**

### CODE REVIEW

File: core/vm/allow_list.go [hunk 1]
```diff
@@ -17,9 +17,38 @@ const (
 
 type allowListPrecompile struct{}
 
+var (
+	enabledKey = common.Hash{}
```
Comment (patrick-ogrady) [inline core/vm/allow_list.go:21]: TODO: is setting a key of `0x000` the same as deleting something?

File: core/vm/allow_list.go [hunk 2]
```diff
@@ -17,9 +17,38 @@ const (
...
+)
+
+type EnabledStatus common.Hash
```
Comment (patrick-ogrady) [inline core/vm/allow_list.go:24]: Pick a new word here instead of `EnabledStatus`...maybe `Activated`

File: core/genesis.go [hunk 3]
```diff
@@ -272,10 +272,19 @@ func (g *Genesis) ToBlock(db ethdb.Database) *types.Block {
 		)
 	}
 	// Configure each address in [ContractAllowListAdmin] to be an admin for the allow list precompile
+	var setAllowList bool
```
Comment (patrick-ogrady) [inline core/genesis.go:275]: nit: `enableAllowList`

File: core/genesis.go [hunk 4]
```diff
@@ -272,10 +272,19 @@ func (g *Genesis) ToBlock(db ethdb.Database) *types.Block {
...
 	for _, addr := range g.ContractAllowListAdmin {
 		if err := vm.SetAllowListStatus(statedb, addr, vm.Admin); err != nil {
 			panic(fmt.Errorf("set allow list errored during genesis configuration: %w", err))
 		}
+		setAllowList = true
+	}
+	// If an admin was added to the allow list, make sure to enable it
+	if setAllowList {
+		// TODO: change naming
+		if err := vm.SetAllowListEnabled(statedb, vm.Enabled); err != nil {
```
Comment (patrick-ogrady) [inline core/genesis.go:285]: nit: `SetAllowListMode`

File: core/genesis.go [hunk 5]
```diff
@@ -272,10 +272,19 @@ func (g *Genesis) ToBlock(db ethdb.Database) *types.Block {
 		)
 	}
 	// Configure each address in [ContractAllowListAdmin] to be an admin for the allow list precompile
+	var setAllowList bool
 	for _, addr := range g.ContractAllowListAdmin {
 		if err := vm.SetAllowListStatus(statedb, addr, vm.Admin); err != nil {
```
Comment (patrick-ogrady) [inline core/genesis.go:277]: nit: `SetAllowListPermissions`

File: core/vm/allow_list.go (hunk 1 above)
Comment (patrick-ogrady) [reply core/vm/allow_list.go:21]: nit: `statusKey`

File: core/vm/allow_list.go [hunk 6]
```diff
@@ -138,8 +180,16 @@ func (al *allowListPrecompile) Run(evm *EVM, caller ContractRef, addr common.Add
 		return nil, remainingGas, ErrExecutionReverted
 	}
 
-	if err := SetAllowListStatus(evm.StateDB, address, status); err != nil {
-		return nil, remainingGas, err
+	// TODO: make this WAY less gross
```
Comment (patrick-ogrady) [inline core/vm/allow_list.go:183]: Just have `UnpackModifyAllowList` return raw `common.Hash,common.Hash`

File: core/vm/allow_list.go (hunk 1 above)
Comment (aaronbuchwald) [reply core/vm/allow_list.go:21]: Yes: https://github.com/ava-labs/subnet-evm/blob/master/core/state/state_object.go#L363

Comment (patrick-ogrady) [reply core/vm/allow_list.go:21]: That link refers to the value, not the key itself.

File: core/vm/evm.go [hunk 7]
```diff
@@ -452,11 +452,13 @@ func (evm *EVM) create(caller ContractRef, codeAndHash *codeAndHash, gas uint64,
 	if evm.StateDB.GetNonce(address) != 0 || (contractHash != (common.Hash{}) && contractHash != emptyCodeHash) {
 		return nil, common.Address{}, 0, ErrContractAddressCollision
 	}
-	// Check that [tx.origin] has the right to deploy contracts
-	modifyStatus := getAllowListStatus(evm.StateDB, evm.Origin)
-	if !modifyStatus.HasDeployPrivileges() {
-		log.Info("origin is not authorized to deploy a contract", "origin", evm.Origin, "status", modifyStatus)
-		return nil, common.Address{}, 0, fmt.Errorf("origin %s is not authorized to deploy a contract", evm.Origin)
+	if getAllowListEnabled(evm.StateDB) {
```
Comment (aaronbuchwald) [inline core/vm/evm.go:455]: I think we should use `chainConfig` here instead to decide whether or not this should be treated as enabled to follow the existing pattern. This should be easier to maintain for optional stateful precompiles going forward.

**Comment by patrick-ogrady** [comment]
closing for #25

**Pull request closed without merging**